test/grouping.jl

```diff
@@ -3577,4 +3577,39 @@ end
     end
 end
 
+@testset "result eltype widening from different tasks" begin
+    if VERSION < v"1.5"
+        Base.convert(::Type{Union{Missing, Nothing, Float64}}, x::Int) = float(x)
+    end
+    Random.seed!(1)
+    for y in (Any[1, missing, missing, 2, 4],
+              Any[1, missing, nothing, 2.1, 'a'],
+              Any[1, 1, missing, 1, nothing, 1, 2.1, 1, 'a'],
+              Any[1, 2, 3, 4, 5, 6, 2.1, missing, 'a'],
+              Any[1, 2, 3.1, 4, 5, 6, 2.1, missing, 'a']),
+        x in (1:length(y), rand(1:2, length(y)), rand(1:3, length(y)))
+        df = DataFrame(x=x, y=y)
+        gd = groupby(df, :x)
+        res = combine(gd, :y => (y -> y[1]) => :y)
+        # sleep ensures one task will widen the result after the other is done,
+        # so that data has to be copied at the end
+        @test res ≅
+              combine(gd, [:x, :y] => ((x, y) -> (sleep((x == [5])/10); y[1])) => :y) ≅
+              combine(gd, [:x, :y] => ((x, y) -> (sleep(x[1]/100); y[1])) => :y) ≅
+              combine(gd, [:x, :y] => ((x, y) -> (sleep(rand()/10); y[1])) => :y)
+
+        if df.x == 1:nrow(df)
+            @test res ≅ df
+        end
+
+        res = combine(gd, :y => (y -> (y1=y[1], y2=last(y))) => AsTable)
+        # sleep ensures one task will widen the result after the other is done,
+        # so that data has to be copied at the end
+        @test res ≅
+              combine(gd, [:x, :y] => ((x, y) -> (sleep((x == [5])/10); (y1=y[1], y2=last(y)))) => AsTable) ≅
+              combine(gd, [:x, :y] => ((x, y) -> (sleep(x[1]/100); (y1=y[1], y2=last(y)))) => AsTable) ≅
+              combine(gd, [:x, :y] => ((x, y) -> (sleep(rand()/10); (y1=y[1], y2=last(y)))) => AsTable)
+    end
```

Comment:
some more tests, especially when multiple transformations are done in the same function would be useful (to check if there is no interference).

Reply:
https://github.com/JuliaData/DataFrames.jl/pull/2574 added lots of tests for multiple operations, and here we're testing type widening, which should happens independently for each transformation. What kind of interference do you think could happen?

Reply:
I have commented above on one such case (several transformations run in parallel that themselves run in parallel).

Reply:
Yes, but even though that should affect performance, why would it change results?

Reply:
I think it does not with the current implementation. But I would add such test now, so that in the future when we get the code even more complex we are covering these cases. And as you know multithreading is very difficult to reason about.